File: app/src/main/java/net/tlalka/android/fiszki/view/adapters/TestsAdapter.java
package net.tlalka.android.fiszki.view.adapters;

import java.util.List;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import net.tlalka.android.fiszki.R;
import net.tlalka.android.fiszki.model.entities.Lesson;

public class TestsAdapter extends AbstractAdapter<Lesson> {

    public TestsAdapter(Context context, List<Lesson> elements) {
        super(context, elements);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup viewGroup) {
        if (convertView == null) {
            convertView = super.getLayoutInflater().inflate(R.layout.test_list_item, viewGroup, false);
            convertView.setTag(new ViewHolderPattern(convertView));
        }

        Lesson lesson = super.getItem(position);

        ViewHolderPattern viewHolderPattern = (ViewHolderPattern) convertView.getTag();
        viewHolderPattern.score.setText(getString(R.string.test_score_value, lesson.getScore()));
        viewHolderPattern.name.setText(getString(R.string.list_item, position + 1, lesson.getName()));
        viewHolderPattern.desc.setText(lesson.getLevelType().name());

        return convertView;
    }

    private static class ViewHolderPattern {
        public ImageView icon;
        public TextView score;
        public TextView name;
        public TextView desc;

        public ViewHolderPattern(View view) {
            this.icon = (ImageView) view.findViewById(R.id.test_list_icon);
            this.score = (TextView) view.findViewById(R.id.test_list_score);
            this.name = (TextView) view.findViewById(R.id.test_list_name);
            this.desc = (TextView) view.findViewById(R.id.test_list_desc);
        }
    }
}

File: app/src/main/java/net/tlalka/android/fiszki/view/adapters/TestsAdapter.kt
package net.tlalka.android.fiszki.view.adapters

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import net.tlalka.android.fiszki.R
import net.tlalka.android.fiszki.model.entities.Lesson

class TestsAdapter(context: Context, elements: List<Lesson>) : AbstractAdapter<Lesson>(context, elements) {

    override fun getView(position: Int, convertView: View?, viewGroup: ViewGroup?): View {
        val itemView = super.getItemView(convertView, viewGroup, R.layout.test_list_item)
        itemView.tag = itemView.tag ?: ViewHolderPattern(itemView)

        val lesson = super.getItem(position)
        val viewHolderPattern = itemView.tag as ViewHolderPattern

        viewHolderPattern.score.text = getString(R.string.test_score_value, lesson.score)
        viewHolderPattern.name.text = getString(R.string.list_item, position + 1, lesson.name)
        viewHolderPattern.desc.text = lesson.levelType.name

        return itemView
    }

    private class ViewHolderPattern(view: View) {
        val icon: ImageView = view.findViewById<View>(R.id.test_list_icon) as ImageView
        val score: TextView = view.findViewById<View>(R.id.test_list_score) as TextView
        val name: TextView = view.findViewById<View>(R.id.test_list_name) as TextView
        val desc: TextView = view.findViewById<View>(R.id.test_list_desc) as TextView
    }
}
